### PR TITLE
Lrtests

### PR DESCRIPTION
**Pull request by ibralibya**
Added tests for LR, modified evaluator tests

### CODE REVIEW

File: tests/evaluator_tests.py [hunk 1]
```diff
@@ -57,12 +57,12 @@ def runTest(self):
         total_ratings = numpy.count_nonzero(self.ratings_matrix)
 
         # ensure that each fold has 1/k of the total ratings
-        k_inverse = (1 / self.k_folds)
-        self.assertEqual(k_inverse, numpy.count_nonzero(test1) / total_ratings)
+        k_inverse = '%.1f' % numpy.round((1 / self.k_folds), 2)
+        self.assertEqual(k_inverse, '%.1f' % numpy.round(numpy.count_nonzero(test1) / total_ratings, 1))
```
Comment (mostafa-mahmoud) [inline tests/evaluator_tests.py]: For these 3 assertions, to assert that 2 floats a and b are equal (or almost equal) it's better to make it like assertTrue(abs(a - b) < EPS) # Where EPS is a small value like 1e-6 (I used this usually) or 1e-8
it's a bit hacky to compare the string representations.

File: tests/evaluator_tests.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ def setUp(self):
         """
         Setup method that is called at the beginning of each test.
         """
-        self.documents, self.users = 18, 10
+        self.documents, self.users = 300, 20
```
Comment (mostafa-mahmoud) [inline tests/evaluator_tests.py]: Why these values? We don't have such number of articles.

File: tests/linear_regression_tests.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        cf = CollaborativeFiltering(self.initializer, self.evaluator, self.hyperparameters,
+                                    self.options, load_matrices=False, is_hybrid=False)
+
+        cf.train()
+        ratings = cf.get_ratings()
+
+        train_data = cf.train_data
+        test_data = cf.test_data
+        cf_predictions = cf.get_predictions()
+
```
Comment (mostafa-mahmoud) [inline tests/linear_regression_tests.py]: remove extra empty lines (lint)

File: tests/linear_regression_tests.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        recall_without_lr = self.evaluator.calculate_recall(self.ratings_matrix, cf_predictions)
+
+        mock_bad_predictions = (numpy.random.randint(2, size=(self.users, self.documents)))
+
+        linearRegressor = LinearRegression(train_data, test_data, mock_bad_predictions, cf_predictions)
```
Comment (mostafa-mahmoud) [inline tests/linear_regression_tests.py]: variable name convention [a-z_0-9]+ (No capital letters, and use underscores) (lint)

File: tests/linear_regression_tests.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        ratings = cf.get_ratings()
+
+        train_data = cf.train_data
+        test_data = cf.test_data
+        cf_predictions = cf.get_predictions()
+
+
+        recall_without_lr = self.evaluator.calculate_recall(self.ratings_matrix, cf_predictions)
+
+        mock_bad_predictions = (numpy.random.randint(2, size=(self.users, self.documents)))
```
Comment (mostafa-mahmoud) [inline tests/linear_regression_tests.py]: no need for extra parentheses (lint)

File: tests/linear_regression_tests.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        self.assertEquals(linearRegressor.flat_item_based_ratings.shape[0], self.users * self.documents)
+        self.assertEquals(linearRegressor.flat_collaborative_ratings.shape[0], self.users * self.documents)
+        self.assertEquals(linearRegressor.flat_train_labels.shape[0], self.users * self.documents)
+        self.assertEquals(linearRegressor.flat_test_labels.shape[0], self.users * self.documents)
+
+        # Because our predictions from the second recommender are random there will be a lot of bad predictions
+        # recall should lower significantly after LR.
+
+        recall_with_lr = self.evaluator.calculate_recall(self.ratings_matrix, linearRegressor.train())
+        self.assertTrue(recall_with_lr < recall_without_lr)
```
Comment (mostafa-mahmoud) [inline tests/linear_regression_tests.py]: This test looks good, we can have an additional test that can make a regression between 2 linearly independent 2d-vectors, something like LR( <2, 5>, T, <1, 0> , <0, 1>) Where T is a test_data; this should make a regression factors of 2, 5

So it wouldn't just be dependent on recommendations.

File: tests/linear_regression_tests.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        def mock_get_ratings_matrix(self=None):
+            return [[int(not bool((article + user) % 3)) for article in range(documents_cnt)]
+                    for user in range(users_cnt)]
+        self.ratings_matrix = numpy.array(mock_get_ratings_matrix())
+        self.evaluator = Evaluator(self.ratings_matrix)
+        setattr(DataParser, "get_ratings_matrix", mock_get_ratings_matrix)
+
+
+
+class TestALS(TestcaseBase):
```
Comment (mostafa-mahmoud) [inline tests/linear_regression_tests.py]: Rename test

File: tests/linear_regression_tests.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        self.options = {'k_folds': self.k_folds, 'n_iterations': self.n_iterations}
+        self.initializer = ModelInitializer(self.hyperparameters.copy(), self.n_iterations)
+
+        def mock_get_ratings_matrix(self=None):
+            return [[int(not bool((article + user) % 3)) for article in range(documents_cnt)]
+                    for user in range(users_cnt)]
+        self.ratings_matrix = numpy.array(mock_get_ratings_matrix())
+        self.evaluator = Evaluator(self.ratings_matrix)
+        setattr(DataParser, "get_ratings_matrix", mock_get_ratings_matrix)
+
```
Comment (mostafa-mahmoud) [inline tests/linear_regression_tests.py]: Remove extra line (lint)

File: tests/evaluator_tests.py [hunk 9]
```diff
@@ -85,32 +85,31 @@ def runTest(self):
...
 
         # removing all top hits, should yield ndcg of 0 as number of recs is 1.
         for i in range(0, self.users):
-            evaluator.ratings[i, self.predictions[i].nonzero()[0]] = 0
+            evaluator.ratings[i] = 0
         ndcg = evaluator.calculate_ndcg(self.n_recommendations, self.predictions,
-                                        self.ratings_matrix, self.test_data)
+                                        evaluator.ratings, self.test_data)
 
         self.assertEqual(0.0, ndcg)
```
Comment (mostafa-mahmoud) [inline tests/evaluator_tests.py]: Was this test fixed? that is the that was failing

File: tests/evaluator_tests.py (hunk 2 above)
Comment (ibralibya) [inline tests/evaluator_tests.py]: done

File: tests/evaluator_tests.py (hunk 1 above)
Comment (ibralibya) [inline tests/evaluator_tests.py]: done

File: tests/evaluator_tests.py (hunk 9 above)
Comment (ibralibya) [inline tests/evaluator_tests.py]: yes it was

File: tests/linear_regression_tests.py (hunk 6 above)
Comment (ibralibya) [inline tests/linear_regression_tests.py]: Test implemented, but it's not achieving required results

Comment (ibralibya) [inline tests/linear_regression_tests.py]: @ndomar any idea why omar?

Comment (ibralibya) [inline tests/linear_regression_tests.py]: Test implemented, issue was with vector dimensionality

**Pull request merged**